aten/src/ATen/native/Linear.cpp

```diff
@@ -41,6 +41,23 @@ Tensor linear(const Tensor& input, const Tensor& weight, const c10::optional<Ten
   return output;
 }
 
+Tensor& linear_out(const Tensor& input, const Tensor& weight, const c10::optional<Tensor>& bias_opt, Tensor& output) {
+  // See [Note: hacky wrapper removal for optional tensor]
+  auto bias = bias_opt.has_value()
+              ? c10::MaybeOwned<Tensor>::borrowed(*bias_opt)
+              : c10::MaybeOwned<Tensor>::owned(c10::in_place);
+
+  if (input.dim() == 2 && bias->defined()) {
+    // Fused op is marginally faster.
+    return at::addmm_out(output, *bias, input, weight.t());
+  }
+  output = at::matmul_out(output, input, weight.t());
```

Comment:
No need to assign the return value to `output` here.

Reply:
yea I saw confused by this (because obv the `output` arg is a ref to memory that gets mutated) but I saw this pattern elsewhere. I thought maybe it might be some kind of convention. will remove.

Reply:
I have a quick question here. What happens if we pass an MKLDNN tensor in here?
The functional linear has this special case:

```cpp
Tensor linear(const Tensor& input, const Tensor& weight, const c10::optional<Tensor>& bias_opt) {
//...
  if (input.is_mkldnn()) {
    return at::mkldnn_linear(input, weight, *bias);
  }
  ```

Reply:
this is an important question. I couldn't figure out how to annotate the schema in `native_functions.yaml` in order to prevent that from happening. For now I'll add

`TORCH_CHECK(!input.is_mkldnn(), "linear doesn't support out for MKLDNN tensors");`